pkg/cmd/experimental/syncgroups/groupsyncer_test.go

```diff
@@ -77,54 +170,94 @@ var Group1Members []*ldap.Entry = []*ldap.Entry{Member1, Member2}
-func TestSync(t *testing.T) {
-	testGroupLister := TestGroupLister{
-		GroupUIDs: []string{Group1UID, Group2UID, Group3UID},
+// TestGoodSync ensures that data is exchanged and rearranged correctly during the sync process.
+func TestGoodSync(t *testing.T) {
+	testGroupSyncer, tc := newTestSyncer()
+	_, errs := testGroupSyncer.Sync()
+	for _, err := range errs {
+		t.Errorf("unexpected sync error: %v", err)
 	}
-	testGroupMemberExtractor := TestGroupMemberExtractor{
-		MemberMapping: map[string][]*ldap.Entry{
-			Group1UID: Group1Members,
-			Group2UID: Group2Members,
-			Group3UID: Group3Members,
-		},
+
+	checkClientForGroups(tc, newDefaultOpenShiftGroups(testGroupSyncer.Host), t)
+}
+
+func TestListFails(t *testing.T) {
+	testGroupSyncer, _ := newTestSyncer()
+	testGroupSyncer.GroupLister.(*TestGroupLister).err = errors.New("error during listing")
+
+	groups, errs := testGroupSyncer.Sync()
+	if len(errs) != 1 {
+		t.Errorf("unexpected sync error: %v", errs)
+
+	} else if errs[0] != testGroupSyncer.GroupLister.(*TestGroupLister).err {
+		t.Errorf("unexpected sync error: %v", errs)
 	}
-	testUserNameMapper := TestUserNameMapper{
-		NameAttributes: []string{UserNameAttribute},
+
+	if groups != nil {
+		t.Errorf("unexpected groups %v", groups)
 	}
-	testGroupNameMapper := TestGroupNameMapper{
-		NameMapping: map[string]string{
-			Group1UID: "os" + Group1UID,
-			Group2UID: "os" + Group2UID,
-			Group3UID: "os" + Group3UID,
-		},
+}
+
+func TestMissingLDAPGroupUIDMapping(t *testing.T) {
+	testGroupSyncer, tc := newTestSyncer()
+	testGroupSyncer.GroupLister.(*TestGroupLister).GroupUIDs = append(testGroupSyncer.GroupLister.(*TestGroupLister).GroupUIDs, "ldapgroupwithnouid")
+
+	_, errs := testGroupSyncer.Sync()
+	if len(errs) != 1 {
+		t.Errorf("unexpected sync error: %v", errs)
+
+	} else if e, a := "no members found for group: ldapgroupwithnouid", errs[0].Error(); e != a {
+		t.Errorf("expected %v, got %v", e, a)
 	}
-	testGroupClient := TestGroupClient{
-		Storage: make(map[string]*userapi.Group),
+
+	checkClientForGroups(tc, newDefaultOpenShiftGroups(testGroupSyncer.Host), t)
+}
+
+func checkClientForGroups(tc *testclient.Fake, expectedGroups []*userapi.Group, t *testing.T) {
+	actualGroups := extractActualGroups(tc)
+
+	for _, expectedGroup := range expectedGroups {
+		if !groupExists(actualGroups, expectedGroup) {
+			t.Errorf("did not find %v, got %v", expectedGroup, actualGroups)
+		}
 	}
-	testHost := "test.host:port"
+}
 
-	testGroupSyncer := LDAPGroupSyncer{
-		GroupLister:          &testGroupLister,
-		GroupMemberExtractor: &testGroupMemberExtractor,
-		UserNameMapper:       &testUserNameMapper,
-		GroupNameMapper:      &testGroupNameMapper,
-		GroupClient:          &testGroupClient,
-		Host:                 testHost,
-		SyncExisting:         false,
+func groupExists(haystack []*userapi.Group, needle *userapi.Group) bool {
+	for _, actual := range haystack {
+		t := *actual
+		actualGroup := &t
```

Comment:
> What is this pointer referencing-dereferencing doing? Can you rename variables here to make it more obvious?

It makes a copy of the data and then points back to it so that you can change the copy without changing the original.

This one in particular is bugged because the map it manipulates is actually a pointer which is pulled into the shallow copy.   Updated.